src/components/NotesMindMap/NotesMindMapReducer.js

```diff
@@ -49,59 +48,65 @@ export const notesMindMapReducer = (state = defaultState, { type, data }) => {
-    noteToUpdate.name = updatedNote.name;
-    return newState;
+    let notes = [ ...state.notes ]
+    notes = notes.filter(note => note.id !== updatedNote.id)
+    notes.push({ id: updatedNote.id, label: updatedNote.name, name: updatedNote.name, isNote: updatedNote.isNote });
+    return { ...state, notes };
   };
 
   const handleCreateNoteSuccessAction = () => {
     const newNote = data;
-    const newState = cloneTreeInState();
-    const parentNote = tree(newState.rootNote).find(note => newNote.parent.id === note.id);
-    parentNote.children.push(newNote);
-    return newState;
+    let notes = [ ...state.notes ]
+    let edges = [ ...state.edges ]
+    notes.push({ id: newNote.id, name: newNote.name, label: newNote.name, isNote: newNote.isNote, group: 'children' });
+    edges.push({ from: newNote.parent.id, to: newNote.id });
+    notes = addGroupTagToNote(notes, state.selectedNote.id)
+      return { ...state, notes, edges };
   };
 
   const handleDeleteNoteRequestSuccessAction = () => {
     const noteToDelete = data;
-    const newState = cloneTreeInState();
-    newState.selectedNote = noteToDelete.parent
-    noteToDelete.parent.children = noteToDelete.parent.children.filter(
-      child => child.id !== noteToDelete.id
-    );
-    return newState;
+    let notes = [ ...state.notes ]
+    let edges = [ ...state.edges ]
+    let graph = removeNoteFromGraph(notes, edges, noteToDelete)
+    notes = graph.notes
+    edges = graph.edges
+    return { ...state, notes, edges };
   };
 
   const handleChangeParentNoteRequestSuccess = () => {
-    const { rootNote } = state;
     const noteId = data.noteId;
     const newParentId = data.newParentId;
-    const targetNote = tree(rootNote).find(node => node.id === noteId);
-    const newParent = tree(rootNote).find(node => node.id === newParentId);
-    const newState = cloneTreeInState();
-
-    newParent.children.push(targetNote);
+    let edges = [ ...state.edges ]
+    let notes = [ ...state.notes ]
+    const oldParent = edges.find(edge => edge.to === noteId);
+    edges = edges.filter(edge => edge.to !== noteId)
+    edges.push({ from: newParentId, to: noteId });
 
-    targetNote.parent.children = targetNote.parent.children.filter(
-      child => child.id !== targetNote.id
-    );
+    notes = updateGroupOfOldParent(notes, edges, newParentId, oldParent.from)
     return {
-      ...newState,
+      ...state,
       isChangeParentModeActive: false,
       showNoteNameEditor: false,
+      edges, notes
     };
   }
 
   const handleChangeSelectedNoteAction = () => {
-    const newState = cloneTreeInState();
-    newState.selectedNote = data;
-    return newState;
+    const newState = { ...state };
+    newState.selectedNote = data.note;
+    return { ...state, selectedNote: data.note };
+  }
+
+  const addRootToGraph = () => {
+    let notes = [ ...state.notes ]
+    notes.push({ id: data.id, label: data.name });
```

Comment:
In other places you push notes with more properties:
```
{ id: updatedNote.id, label: updatedNote.name, name: updatedNote.name, isNote: updatedNote.isNote }
```
Why does the root note has less properties?